### PR TITLE
Added `delete_namespace` and `delete_project` functions

### PR DESCRIPTION
**Pull request by ilongin**
Up until now we didn't have the way to remove namespace or project. This PR adds 2 functions: `delete_project` and `delete_namespace`

There are some rules:
1. Cannot delete default namespace or project
2. Cannot delete system namespace
3. Cannot delete listing project
4. Cannot delete non empty project (has datasets in it)
5. Cannot delete non empty namespace (has projects in it)

## Summary by Sourcery

Enable removal of projects and namespaces with corresponding metastore support, error handling, and comprehensive tests.

New Features:
- Add delete_project function to remove a project with safety checks
- Add delete_namespace function to remove a namespace with safety checks

Enhancements:
- Extend metastore with remove_project, remove_namespace, count_datasets, is_default_project, and is_listing_project methods

Tests:
- Add unit tests for delete_project covering success, not found, listing/default, and non-empty scenarios
- Add unit tests for delete_namespace covering success, not found, system/default, and non-empty scenarios

Chores:
- Define ProjectDeleteNotAllowedError and NamespaceDeleteNotAllowedError error classes
- Export delete_project in the public API

### CODE REVIEW

**Comment by shcheklein** [comment]
Ci fails?

File: src/datachain/data_storage/metastore.py [hunk 1]
```diff
@@ -190,10 +195,22 @@ def get_project(
         It also creates project if not found and create flag is set to True.
         """
 
+    @abstractmethod
+    def is_default_project(self, project_name: str, namespace_name: str) -> bool:
+        """Checks if some project in namespace is default"""
+
+    @abstractmethod
+    def is_listing_project(self, project_name: str, namespace_name: str) -> bool:
```
Comment (shcheklein) [inline src/datachain/data_storage/metastore.py:206]: Can we prevent it in the db level as well? Have a flag for all special ovjects. To make it more reliable and not easy to hack

File: src/datachain/lib/namespaces.py [hunk 2]
```diff
@@ -71,3 +74,49 @@ def ls(session: Optional[Session] = None) -> list[Namespace]:
...
+            as these cannot be removed.
+
+    Parameters:
+        name : The name of the namespace.
+        session : Session to use for getting project.
+
+    Example:
+        ```py
+        import datachain as dc
+        from datachain.lib.namespace import delete as delete_namespace
```
Comment (shcheklein) [inline src/datachain/lib/namespaces.py:97]: This looks bad. Like API is not named good and we have to rename it

File: src/datachain/lib/projects.py [hunk 3]
```diff
@@ -86,3 +86,50 @@ def ls(
...
+        from datachain.lib.projects import delete as delete_project
+        delete_project("my-project", "local")
+        ```
+    """
+    session = Session.get(session)
+    metastore = session.catalog.metastore
+
+    project = metastore.get_project(name, namespace)
+
+    if metastore.is_listing_project(name, namespace):
```
Comment (shcheklein) [inline src/datachain/lib/projects.py:117]: This checks should be also happening on the API side + on the db level.

File: src/datachain/__init__.py [hunk 4]
```diff
@@ -74,6 +76,8 @@
     "create_project",
     "datasets",
     "delete_dataset",
+    "delete_namespace",
+    "delete_project",
```
Comment (dmpetrov) [inline src/datachain/__init__.py]: Let's don't expose project in user's api. Instead you can do something like:

```python
def delete_namespace(name: str):
    parts = name.split(".")
    if len(parts) == 1:
        remove_namespace(parts[0])
    elif len(parts) == 2:
        delete_project(parts[1])
    else:
        raise ValueError(f"Invalid namespace format: {name}. Expected 'namespace' or 'ns1.ns2'.")
```

**Review by dmpetrov — CHANGES_REQUESTED** [review]
LG but there is a request to change public API

File: src/datachain/data_storage/metastore.py [hunk 5]
```diff
@@ -141,6 +142,10 @@ def create_namespace(
     def get_namespace(self, name: str, conn=None) -> Namespace:
         """Gets a single namespace by name"""
 
+    @abstractmethod
+    def remove_namespace(self, namespace_id: int, conn=None) -> None:
+        """Removes a single namespace by id"""
```
Comment (dreadatour) [inline src/datachain/data_storage/metastore.py:149]: Just a heads-up: if we add new methods to the metastore in the Datachain project, we’ll also need a corresponding Studio PR. Otherwise, merging the Datachain PR first could potentially break the Studio release.

File: src/datachain/data_storage/metastore.py [hunk 6]
```diff
@@ -735,6 +756,11 @@ def create_namespace(
 
         return self.get_namespace(name)
 
+    def remove_namespace(self, namespace_id: int, conn=None) -> None:
+        n = self._namespaces
+        with self.db.transaction():
+            self.db.execute(self._namespaces_delete().where(n.c.id == namespace_id))
```
Comment (dreadatour) [inline src/datachain/data_storage/metastore.py:779]: I am not sure, just wonder, should we add another check here (so no datasets exists within this namespace), or is it enough to have this check on a higher level? May be consider to move the check here and raise an exceptions (like "default namespace can not be deleted" or "namespace with datasets can not be deleted) to be on the safe side? 🤔

File: src/datachain/data_storage/metastore.py [hunk 7]
```diff
@@ -858,6 +884,11 @@ def get_project_by_id(self, project_id: int, conn=None) -> Project:
             raise ProjectNotFoundError(f"Project with id {project_id} not found.")
         return self.project_class.parse(*rows[0])
 
+    def remove_project(self, project_id: int, conn=None) -> None:
+        p = self._projects
+        with self.db.transaction():
+            self.db.execute(self._projects_delete().where(p.c.id == project_id))
```
Comment (dreadatour) [inline src/datachain/data_storage/metastore.py:912]: Same question here with checks on this level 🤔

File: src/datachain/data_storage/metastore.py [hunk 8]
```diff
@@ -1198,6 +1228,16 @@ def list_datasets(
             query = query.where(d.c.project_id == project_id)
         yield from self._parse_dataset_list(self.db.execute(query))
 
+    def count_datasets(self, project_id: Optional[int] = None) -> int:
+        d = self._datasets
+        query = self._base_list_datasets_query()
```
Comment (dreadatour) [inline src/datachain/data_storage/metastore.py]: This is not optimal IMO, since it will JOIN all datasets versions. Should we have simpler `COUNT` query here to count datasets only?

File: src/datachain/lib/namespaces.py [hunk 9]
```diff
@@ -71,3 +74,49 @@ def ls(session: Optional[Session] = None) -> list[Namespace]:
...
+        raise NamespaceDeleteNotAllowedError(
+            f"Namespace {metastore.system_namespace_name} cannot be removed"
+        )
+
+    if name == metastore.default_namespace_name:
+        raise NamespaceDeleteNotAllowedError(
+            f"Namespace {metastore.default_namespace_name} cannot be removed"
+        )
+
+    projects = metastore.list_projects(namespace.id)
```
Comment (dreadatour) [inline src/datachain/lib/namespaces.py]: May be add `count_projects` or `has_projects` method to do not fetch all the projects (similar to `count_datasets`)?
Otherwise in Studio this will pass all the projects objects via HTTP API.

Also for the consistency I think we should stick to either `has_datasets` and `has_projects` or `count_datasets` and `count_projects`.

**Comment by shcheklein** [comment]
@ilongin what is the progress here?

File: src/datachain/data_storage/metastore.py (hunk 1 above)
Comment (ilongin) [reply src/datachain/data_storage/metastore.py:206]: We could,  yes, but I would like to do it in a followup as it requires more changes, DB migration etc.

File: src/datachain/lib/namespaces.py (hunk 2 above)
Comment (ilongin) [reply src/datachain/lib/namespaces.py:97]: This was just a bad example. Correct one is:
```python
import datachain as dc
dc.delete_project(...)
```

Comment (shcheklein) [reply src/datachain/lib/namespaces.py:97]: Then we need to update the file?

File: src/datachain/__init__.py (hunk 4 above)
Comment (ilongin) [reply src/datachain/__init__.py]: I didn't want to do this here as there is an open issue about this to change it everywhere (and be consistent) https://github.com/iterative/datachain/issues/1302  ..

File: src/datachain/lib/namespaces.py (hunk 2 above)
Comment (ilongin) [reply src/datachain/lib/namespaces.py:97]: Yes, of course .. updated

File: src/datachain/data_storage/metastore.py (hunk 5 above)
Comment (ilongin) [reply src/datachain/data_storage/metastore.py:149]: Yea, I again totally forgot about it .. thanks for the comment. Added Studio PR

File: src/datachain/data_storage/metastore.py (hunk 6 above)
Comment (ilongin) [reply src/datachain/data_storage/metastore.py:779]: I don't think that check should be added here. It's a higher level business logic check and metastore, IMO should be de-coupled of that .. it should be just simple DB layer for other parts to use and that's it. For example, maybe we will have some management command to actually remove those default namespaces at some point and we want to be able to use metastore.

File: src/datachain/data_storage/metastore.py (hunk 7 above)
Comment (ilongin) [reply src/datachain/data_storage/metastore.py:912]: Same answer as above.... I personally wouldn't put these kind of things here. It's business logic

File: src/datachain/data_storage/metastore.py (hunk 8 above)
Comment (ilongin) [reply src/datachain/data_storage/metastore.py]: Fixed

File: src/datachain/lib/namespaces.py (hunk 9 above)
Comment (ilongin) [reply src/datachain/lib/namespaces.py]: Yes, I agree. I didn't put too much effort into optimization as this functions will not be called often and are working on small number of objects, but I refactored it now to use new count method.

File: src/datachain/data_storage/metastore.py (hunk 6 above)
Comment (dreadatour) [reply src/datachain/data_storage/metastore.py:779]: I believe it not a business logic, but DB level constraints.

Since we don't have foreign keys at the DB level to prevent deletion of relations and if someone will use this `remove_namespace` metastore method to delete namespace, db will be broken — we will have records (datasets) with namespace IDs and no namespace exists in the DB.

Also if we'll want to use `remove_namespace` metastore method from somewhere else, we should implement these same sanity checks in multiple places. It is more practical to move it on a metastore level to do not duplicate the code.

Comment (ilongin) [reply src/datachain/data_storage/metastore.py:779]: Aha, I thought you were referring to removing default / system namespaces and checks for that .. sorry. Anyway, yes, I added additional check for empty namespace / project on removal, that makes sense to me.

File: src/datachain/__init__.py (hunk 4 above)
Comment (dmpetrov) [reply src/datachain/__init__.py]: Not sure I follow 🙂

We’re exposing an API that’ll be removed soon anyway. Skipping it would just mean ~5 extra lines of code.

Or am I missing something?

File: src/datachain/lib/namespaces.py [hunk 10]
```diff
@@ -71,3 +74,49 @@ def ls(session: Optional[Session] = None) -> list[Namespace]:
...
+        raise NamespaceDeleteNotAllowedError(
+            f"Namespace {metastore.default_namespace_name} cannot be removed"
+        )
+
+    num_projects = metastore.count_projects(namespace.id)
+    if num_projects > 0:
+        raise NamespaceDeleteNotAllowedError(
+            f"Namespace cannot be removed. It contains {num_projects} project(s). "
+            "Please remove the project(s) first."
+        )
```
Comment (dreadatour) [inline src/datachain/lib/namespaces.py:126]: We can now remove this check from here.

File: src/datachain/lib/projects.py [hunk 11]
```diff
@@ -86,3 +86,49 @@ def ls(
...
+        raise ProjectDeleteNotAllowedError(
+            f"Project {metastore.default_project_name} cannot be removed"
+        )
+
+    num_datasets = metastore.count_datasets(project.id)
+    if num_datasets > 0:
+        raise ProjectDeleteNotAllowedError(
+            f"Project cannot be removed. It contains {num_datasets} dataset(s). "
+            "Please remove the dataset(s) first."
+        )
```
Comment (dreadatour) [inline src/datachain/lib/projects.py:132]: We can now remove this check from here.

File: src/datachain/__init__.py (hunk 4 above)
Comment (ilongin) [reply src/datachain/__init__.py]: I wanted to be consistent in all of our APIs and change them all together, but ok, we can do this one in correct way right away. I changed it, please approve if you think this is ok now.
